### ISSUE TITLE
strange decay (probably a bug with -decay epoch_only)

### ISSUE THREAD

**Issue by vince62s** [OpenNMT/OpenNMT#264]
command line is start decay epoch 80 and epoch_only

```
[05/06/17 14:47:55 INFO] Epoch 20 ; Iteration 1585/1585 ; Learning rate 1.0000 ; Source tokens/s 5187 ; Perplexity 5.91	
[05/06/17 14:47:59 INFO] Validation perplexity: 5.55	
[05/06/17 14:47:59 INFO] Saving checkpoint to 'exp/model-enfr_epoch20_5.55.t7'...	
[05/06/17 14:48:01 INFO] 	
[05/06/17 14:48:01 INFO] Sampling dataset...	
[05/06/17 14:48:04 INFO] Sampled 100000 instances into 1588 batches.	
[05/06/17 14:48:04 INFO] Importance Sampling - keeping 38376 target vocabs.	
[05/06/17 14:50:04 INFO] Epoch 21 ; Iteration 500/1588 ; Learning rate 1.0000 ; Source tokens/s 5272 ; Perplexity 5.99	
[05/06/17 14:52:09 INFO] Epoch 21 ; Iteration 1000/1588 ; Learning rate 1.0000 ; Source tokens/s 5118 ; Perplexity 5.98	
[05/06/17 14:54:11 INFO] Epoch 21 ; Iteration 1500/1588 ; Learning rate 1.0000 ; Source tokens/s 5227 ; Perplexity 5.92	
[05/06/17 14:54:33 INFO] Epoch 21 ; Iteration 1588/1588 ; Learning rate 1.0000 ; Source tokens/s 5132 ; Perplexity 6.03	
[05/06/17 14:54:37 INFO] Validation perplexity: 5.34	
[05/06/17 14:54:37 INFO] 	
[05/06/17 14:54:37 INFO] Sampling dataset...	
[05/06/17 14:54:41 INFO] Sampled 100000 instances into 1587 batches.	
[05/06/17 14:54:41 INFO] Importance Sampling - keeping 38444 target vocabs.	
[05/06/17 14:56:45 INFO] Epoch 22 ; Iteration 500/1587 ; Learning rate 0.9500 ; Source tokens/s 5157 ; Perplexity 5.82	
[05/06/17 14:58:49 INFO] Epoch 22 ; Iteration 1000/1587 ; Learning rate 0.9500 ; Source tokens/s 5159 ; Perplexity 5.76	
[05/06/17 15:00:53 INFO] Epoch 22 ; Iteration 1500/1587 ; Learning rate 0.9500 ; Source tokens/s 5160 ; Perplexity 5.80	
[05/06/17 15:01:14 INFO] Epoch 22 ; Iteration 1587/1587 ; Learning rate 0.9500 ; Source tokens/s 5086 ; Perplexity 5.51	
[05/06/17 15:01:18 INFO] Validation perplexity: 5.22	
```

**Comment by guillaumekln** [comment]
Strange indeed. Could you share the full list of options set?

**Comment by vince62s** [comment]
if [ $stage -le 3 ]; then
    learning_rate=1
    start_decay_at=80
    end_epoch=130
    learning_rate_decay=0.95
    echo "$0: training public corpus starting, will take a while."
      # train first epoch
if true; then
      th train.lua -layers 2 -rnn_size 500 -brnn -data exp/data-$sl$tl-train.t7 \
      -save_model exp/model-$sl$tl -dropout 0.3 -report_every 500 -word_vec_size 500 \
      -start_epoch 1 -end_epoch $end_epoch -max_batch_size 64 -sample 100000 -sample_tgt_vocab \
      -learning_rate $learning_rate -start_decay_at $start_decay_at -decay epoch_only\
      -learning_rate_decay $learning_rate_decay -save_every_epochs 10 -gpuid $gpuid

**Comment by vince62s** [comment]
another note, before ppl 5.55 it was 5.49 (on epoch 19) in case it may help.

**Comment by vince62s** [comment]
it is related to the above:
again here with a different vocab size only:
```
[05/06/17 19:23:12 INFO] Epoch 17 ; Iteration 1586/1586 ; Learning rate 1.0000 ; Source tokens/s 4592 ; Perplexity 6.64	
[05/06/17 19:23:19 INFO] Validation perplexity: 5.90	
[05/06/17 19:23:19 INFO] 	
[05/06/17 19:23:19 INFO] Sampling dataset...	
[05/06/17 19:23:22 INFO] Sampled 100000 instances into 1587 batches.	
[05/06/17 19:23:22 INFO] Importance Sampling - keeping 46934 target vocabs.	
[05/06/17 19:25:46 INFO] Epoch 18 ; Iteration 500/1587 ; Learning rate 1.0000 ; Source tokens/s 4643 ; Perplexity 6.47	
[05/06/17 19:28:08 INFO] Epoch 18 ; Iteration 1000/1587 ; Learning rate 1.0000 ; Source tokens/s 4567 ; Perplexity 6.45	
[05/06/17 19:30:21 INFO] Epoch 18 ; Iteration 1500/1587 ; Learning rate 1.0000 ; Source tokens/s 4541 ; Perplexity 6.40	
[05/06/17 19:30:46 INFO] Epoch 18 ; Iteration 1587/1587 ; Learning rate 1.0000 ; Source tokens/s 4505 ; Perplexity 6.35	
[05/06/17 19:30:52 INFO] Validation perplexity: 6.12	
[05/06/17 19:30:52 INFO] 	
[05/06/17 19:30:52 INFO] Sampling dataset...	
[05/06/17 19:30:56 INFO] Sampled 100000 instances into 1588 batches.	
[05/06/17 19:30:56 INFO] Importance Sampling - keeping 46775 target vocabs.	
[05/06/17 19:33:17 INFO] Epoch 19 ; Iteration 500/1588 ; Learning rate 1.0000 ; Source tokens/s 4591 ; Perplexity 6.32	
[05/06/17 19:35:37 INFO] Epoch 19 ; Iteration 1000/1588 ; Learning rate 1.0000 ; Source tokens/s 4579 ; Perplexity 6.36	
[05/06/17 19:37:53 INFO] Epoch 19 ; Iteration 1500/1588 ; Learning rate 1.0000 ; Source tokens/s 4569 ; Perplexity 6.23	
[05/06/17 19:38:18 INFO] Epoch 19 ; Iteration 1588/1588 ; Learning rate 1.0000 ; Source tokens/s 4490 ; Perplexity 6.17	
[05/06/17 19:38:24 INFO] Validation perplexity: 5.68	
[05/06/17 19:38:24 INFO] 	
[05/06/17 19:38:24 INFO] Sampling dataset...	
[05/06/17 19:38:28 INFO] Sampled 100000 instances into 1589 batches.	
[05/06/17 19:38:28 INFO] Importance Sampling - keeping 46995 target vocabs.	
[05/06/17 19:40:48 INFO] Epoch 20 ; Iteration 500/1589 ; Learning rate 0.9500 ; Source tokens/s 4565 ; Perplexity 6.17	
[05/06/17 19:43:07 INFO] Epoch 20 ; Iteration 1000/1589 ; Learning rate 0.9500 ; Source tokens/s 4563 ; Perplexity 6.11	
[05/06/17 19:45:27 INFO] Epoch 20 ; Iteration 1500/1589 ; Learning rate 0.9500 ; Source tokens/s 4551 ; Perplexity 6.14	
[05/06/17 19:45:52 INFO] Epoch 20 ; Iteration 1589/1589 ; Learning rate 0.9500 ; Source tokens/s 4510 ; Perplexity 6.28	
[05/06/17 19:45:58 INFO] Validation perplexity: 5.60	
[05/06/17 19:45:58 INFO] Saving checkpoint to 'exp/model-enfr_epoch20_5.60.t7'...	
[05/06/17 19:46:01 INFO] 	
```

**Comment by guillaumekln** [comment]
Sorry, the implemented logic was a bit shaky. I cleaned this up and it should work as expected now.
